scopehal-apps: init at 0-unstable-2024-08-22

bgamari opened the pull request:
## Description of changes

Add [`ngscopeclient`](https://www.ngscopeclient.org/) and its dependency, [`ffts`](https://github.com/anthonix/ffts). As the original `ffts` repository is unmaintained I have opted to use https://github.com/ValveSoftware/ffts.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

bgamari committed:
scopehal-apps: init at unstable-2024-08-22

Co-Authored-By: Hans Baier <foss@hans-baier.de>

bgamari committed:
ffts: init at unstable-2019-03-19

Co-Authored-By: Hans Baier <foss@hans-baier.de>

erdnaxe commented on pkgs/by-name/ff/ffts/package.nix:
```diff
@@ -0,0 +1,42 @@
+    "-Wno-deprecated"
+  ];
+
+  meta = with lib; {
```
```suggestion
  meta = {
```
See https://github.com/NixOS/nixpkgs/issues/208242
You are already using `lib.maintainers` and `lib.platforms` :)

bgamari replied:
Indeed; fixed.

erdnaxe commented on pkgs/by-name/ff/ffts/package.nix:
```diff
@@ -0,0 +1,42 @@
+    license = licenses.mit;
+    maintainers = with lib.maintainers; [
+      thoughtpolice
+      hansfbaier
```
hansfbaier is missing from `maintainers/maintainer-list.nix`. This is making ofborg fail 😢

bgamari replied:
I have dropped @hansfbaier and @thoughtpolice until they confirm that they are willing to maintain.

erdnaxe commented on pkgs/by-name/ff/ffts/package.nix:
```diff
@@ -0,0 +1,42 @@
+    description = "The Fastest Fourier Transform in the South";
+    homepage = "http://anthonix.com/ffts";
+    license = licenses.mit;
+    maintainers = with lib.maintainers; [
```
Why are you not putting `bgamari` as maintainer?
Does `thoughtpolice` and `hansfbaier` want to maintain this package?

bgamari replied:
This came from the original MR, https://github.com/NixOS/nixpkgs/pull/29816. I'll let @hansfbaier and @thoughtpolice confirm.

I will add myself as well.

erdnaxe commented on pkgs/by-name/sc/scopehal-apps/package.nix:
```diff
@@ -0,0 +1,102 @@
+    license = licenses.bsd3;
+    maintainers = with lib.maintainers; [
+      thoughtpolice
+      hansfbaier
```
Same question as for ffts package: does `thoughtpolice` and `hansfbaier` want to maintain this package?

hansfbaier replied:
I put in @thoughtpolice, because seems to be everywhere.
I am not sure if I have the necessary experience in maintaining this.
But I basically might volunteer, but you seem to have much more flight time with doing this.

erdnaxe commented on pkgs/by-name/ff/ffts/package.nix:
```diff
@@ -0,0 +1,42 @@
+
+  meta = with lib; {
+    description = "The Fastest Fourier Transform in the South";
+    homepage = "http://anthonix.com/ffts";
```
This homepage seems dead. Maybe directly link Valve github fork?

erdnaxe commented on pkgs/by-name/sc/scopehal-apps/package.nix:
```diff
@@ -0,0 +1,102 @@
+    ffts
+  ];
+
+  enableParallelBuilding = true;
```
Do we have a rationale for enabling this? Does the compilation takes too long without parallel building?

bgamari replied:
Yes, it is a fairly bulky C++ project. I personally found it very annoying to iterate on this derivation without it.

erdnaxe replied:
CMake already set `enableParallelBuilding = true` by default so it is not necessary.
This line should be removed and it shouldn't impact build time.

erdnaxe reviewed (changes requested):
Thanks for helping package ngscopeclient, it is a really cool piece of software.

erdnaxe commented on pkgs/by-name/sc/scopehal-apps/package.nix:
```diff
@@ -0,0 +1,102 @@
+
+stdenv.mkDerivation rec {
+  pname = "scopehal-apps";
+  version = "e75c0bd1eed4c057bfbbfa6a75a211a88d30c934";
```
```suggestion
  version = "unstable-2024-08-22";
```
See contribution guidelines.

bgamari replied:
Thanks!

erdnaxe commented:
Commits shouldn't say "init at master", but rather match the derivation version (e.g. "init at unstable-2024-01-01").

erdnaxe commented:
> As the original ffts repository is unmaintained I have opted to use https://github.com/ValveSoftware/ffts.

Valve seems to follow https://github.com/linkotec/ffts with just an extra branch for steamaudio. IMHO we should package from `linkotec/ffts` repository.

Also, as ffts has no recent stable releases, you should use "unstable-2019-03-19" as version.

bgamari commented:
Thanks for your review, @erdnaxe!

erdnaxe commented on pkgs/by-name/sc/scopehal-apps/package.nix:
```diff
@@ -0,0 +1,100 @@
+  src = fetchFromGitHub {
+    owner = "ngscopeclient";
+    repo = "scopehal-apps";
+    rev = version;
```
```suggestion
    rev = "e75c0bd1eed4c057bfbbfa6a75a211a88d30c934";
```
else it won't build (`fatal: couldn't find remote ref refs/tags/unstable-2024-08-22`)

erdnaxe commented on pkgs/by-name/sc/scopehal-apps/package.nix:
```diff
@@ -0,0 +1,100 @@
+  ];
+
+  buildInputs = [
+    pkg-config
```
Putting pkg-config in buildInputs breaks cross-compilation.
Please move it to `nativeBuildInputs`.

erdnaxe commented on pkgs/by-name/ff/ffts/package.nix:
```diff
@@ -0,0 +1,41 @@
+  lib,
+  cmake,
+  git,
+  fetchFromGitHub,
```
Please fix styling here, you may use `nixpkgs-fmt` to format the file.
You can also use https://github.com/jtojnar/nixpkgs-hammering as a linter.

erdnaxe commented on pkgs/by-name/ff/ffts/package.nix:
```diff
@@ -0,0 +1,41 @@
+
+  nativeBuildInputs = [
+    cmake
+    git
```
```suggestion
```
Git is not required. The cloning happends before the build, and the build does not have Internet access.
Also remember to remove `git` from the derivation inputs.

erdnaxe commented on pkgs/by-name/ff/ffts/package.nix:
```diff
@@ -0,0 +1,41 @@
+  cmakeFlags = [
+    "-DCURRENT_GIT_VERSION=${lib.substring 0 7 src.rev}"
+    "-DENABLE_SHARED=ON"
+    "-Wno-deprecated"
```
```suggestion
    "-Wno-deprecated"
```
IMHO, we shouldn't remove build warnings.

erdnaxe commented on pkgs/by-name/ff/ffts/package.nix:
```diff
@@ -0,0 +1,41 @@
+  buildInputs = [ ];
+
+  cmakeFlags = [
+    "-DCURRENT_GIT_VERSION=${lib.substring 0 7 src.rev}"
```
```suggestion
```
Unused.

erdnaxe commented on pkgs/by-name/sc/scopehal-apps/package.nix:
```diff
@@ -0,0 +1,100 @@
+  vulkan-tools,
+  glslang,
+  spirv-tools,
+  ffts,
```
Same as ffts, please format with `nixpkgs-fmt`.

erdnaxe commented on pkgs/by-name/sc/scopehal-apps/package.nix:
```diff
@@ -0,0 +1,100 @@
+
+  nativeBuildInputs = [
+    cmake
+    git
```
```suggestion
```
`git` is not needed for the build process.

erdnaxe commented on pkgs/by-name/sc/scopehal-apps/package.nix:
```diff
@@ -0,0 +1,100 @@
+    vulkan-tools
+    spirv-tools
+    glslang
+    shaderc
```
Same as pkg-config. Please move shaderc to `nativeBuildInputs`.

erdnaxe commented on pkgs/by-name/sc/scopehal-apps/package.nix:
```diff
@@ -0,0 +1,100 @@
+    fetchSubmodules = true;
+  };
+
+  nativeBuildInputs = [
```
```suggestion
  strictDeps = true;
  nativeBuildInputs = [
```
By enabling strictDeps, you can make sure `buildInputs` and `nativeBuildInputs` are right.
This will help fix crosscompilation.

erdnaxe commented on pkgs/by-name/sc/scopehal-apps/package.nix:
```diff
@@ -0,0 +1,100 @@
+      --replace '"/share/' '"/../share/'
+  '';
+
+  cmakeFlags = [ "-DCURRENT_GIT_VERSION=${lib.substring 0 7 src.rev}" ];
```
```suggestion
```
This is not needed.

erdnaxe commented on pkgs/by-name/sc/scopehal-apps/package.nix:
```diff
@@ -0,0 +1,100 @@
+{
+  stdenv,
+  breakpointHook,
```
```suggestion
```
unused

erdnaxe reviewed (changes requested):
Are you sure all buildInputs are required to build this package? For example libllvm seems weird as a buildInputs.

bgamari commented:
> Are you sure all buildInputs are required to build this package? For example libllvm seems weird as a buildInputs.

I'll admit, I didn't look as carefully as I should have at the dependencies; they are largely carried over from https://github.com/NixOS/nixpkgs/pull/29816. I have dropped the more-obviously bogus dependencies (`libllvm`, `libdrm`, `elfutils`, `libxcb`, `catch2`, `libxshmfence`, `xcbutilkeysyms`, `systemd`, `git`, and `breakpointHook`).

I believe I have also addressed your remaining feedback. Thanks again for your review!

bgamari commented:
It appears that the format produced by `nix run nixpkgs#nixpkgs-fmt` does not satisfy the linter. I could swear I ran `nix run nixpkgs#nix-fmt` on the earlier iteration but I will try again.

hansfbaier commented:
> libllvm

glscopeclient is GPU accelerated. GPU accelerated software uses libllvm to compile shaders.
Without this the program won't start.

bgamari commented:
@hansfbaier, I believe shader compilation is the responsibility of the OpenGL/Vulkan driver (typically Mesa), not the client. A direct dependency on `libllvm` is not necessary.

hansfbaier commented:
> @hansfbaier, I believe shader compilation is the responsibility of the OpenGL/Vulkan driver (typically Mesa), not the client. A direct dependency on `libllvm` is not necessary.
Try it out. If the program still starts up without that, then it is ok.

bgamari commented:
The linting situation here is rather unfortunate. It appears that the style currently produced by `nixpkgs#nixfmt-classic`  is rejected by the linter. As is the style produced by `nixpkgs#nixpkgs-fmt`. It appears `nixpkgs#nixfmt-rfc-style` is the right incantation.

bgamari commented:
> > @hansfbaier, I believe shader compilation is the responsibility of the OpenGL/Vulkan driver (typically Mesa), not the client. A direct dependency on `libllvm` is not necessary.
> Try it out. If the program still starts up without that, then it is ok.

I have already confirmed that it works as expected.

erdnaxe commented:
aarch64-linux seems to fail due to multiple[ `narrowing conversion of '-1' from 'int' to 'char'` errors](https://stackoverflow.com/questions/31634962/narrowing-conversion-from-int-to-char-inside-for-legal-values-when-cro) in `/build/source/lib/scopehal/base64.cpp`. 

I sent a patch to upstream at https://github.com/ngscopeclient/scopehal/pull/888. We should maybe wait for an upstream fix.

---

On another note, x86 compilation is working great, and I am able to successfully build and run ngscopeclient, thanks!

Aleksanaa commented on pkgs/by-name/ff/ffts/package.nix:
```diff
@@ -0,0 +1,31 @@
+  };
+
+  nativeBuildInputs = [ cmake ];
+  buildInputs = [ ];
```
```suggestion
```

Aleksanaa commented on pkgs/by-name/ff/ffts/package.nix:
```diff
@@ -0,0 +1,31 @@
+
+stdenv.mkDerivation rec {
+  pname = "ffts";
+  version = "unstable-2019-03-19";
```
```suggestion
  version = "0-unstable-2019-03-19";
```

Aleksanaa commented on pkgs/by-name/sc/scopehal-apps/package.nix:
```diff
@@ -0,0 +1,77 @@
+
+stdenv.mkDerivation rec {
+  pname = "scopehal-apps";
+  version = "unstable-2024-08-22";
```
```suggestion
  version = "0-unstable-2024-08-22";
```

Aleksanaa commented on pkgs/by-name/sc/scopehal-apps/package.nix:
```diff
@@ -0,0 +1,77 @@
+    homepage = "https://www.ngscopeclient.org/";
+    license = lib.licenses.bsd3;
+    maintainers = with lib.maintainers; [ bgamari ];
+    platforms = lib.platforms.linux;
```
missing `mainProgram` and `description`

bgamari commented:
Thanks @Aleksanaa !  I have updated the `scopehal-apps` commit and addressed the above suggestions.

bgamari commented:
I believe this is ready to be merged.

bgamari commented:
It is not clear to me why the remaining ofborg job has yet to begin running.

kirillrdy commented on pkgs/by-name/sc/scopehal-apps/package.nix:
```diff
@@ -0,0 +1,79 @@
+  # Targets InitializeSearchPaths
+  postPatch = ''
+    substituteInPlace lib/scopehal/scopehal.cpp \
+      --replace '"/share/' '"/../share/'
```
```suggestion
      --replace--fail '"/share/' '"/../share/'
```

bgamari commented:
Addressed final comment. This is ready for merge.

RossComputerGuy commented:
Result of `nixpkgs-review pr 336646` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>scopehal-apps</li>
  </ul>
</details>
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>ffts</li>
  </ul>
</details>
